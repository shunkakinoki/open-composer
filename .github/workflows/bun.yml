name: Bun
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  workflow_dispatch:
concurrency:
  group: ${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.sha }}
  cancel-in-progress: true
jobs:
  bun-build:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Build
        run: bun run build
  bun-lint:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Check
        run: bun run check
      - name: Format
        run: bun run format
      - name: Lint
        run: bun run lint
      - name: Type Check
        run: bun run type-check
  bun-test:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Test
        run: bun run test
  bun-check:
    if: always()
    needs:
      - bun-build
      - bun-lint
      - bun-test
    runs-on: ubuntu-latest
    timeout-minutes: 3
    steps:
      - name: Alls Green
        uses: re-actors/alls-green@release/v1
        with:
          jobs: ${{ toJSON(needs) }}
