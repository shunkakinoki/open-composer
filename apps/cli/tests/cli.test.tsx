import { describe, expect, it } from "bun:test";
import { spawn } from "node:child_process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const cliPath = join(__dirname, "../src/index.ts");

interface CliResult {
  stdout: string;
  stderr: string;
  code: number | null;
}

const stripAnsi = (value: string): string =>
  value.replace(new RegExp(`${String.fromCharCode(27)}\\[[0-9;]*m`, "g"), "");

const runCli = (args: string[] = []): Promise<CliResult> =>
  new Promise((resolve, reject) => {
    const child = spawn("bun", ["run", cliPath, ...args], {
      stdio: "pipe",
    });

    let stdout = "";
    let stderr = "";

    child.stdout?.on("data", (data) => {
      stdout += data.toString();
    });

    child.stderr?.on("data", (data) => {
      stderr += data.toString();
    });

    child.on("close", (code) => {
      resolve({ stdout, stderr, code });
    });

    child.on("error", (error) => {
      reject(error);
    });
  });

describe("CLI Execution", () => {
  it("prints help text by default", async () => {
    const result = await runCli();
    const stdout = stripAnsi(result.stdout);
    const stderr = stripAnsi(result.stderr);

    expect(result.code).toBe(0);
    expect(stderr).toBe("");
    expect(stdout).toContain("USAGE");
    expect(stdout).toContain("$ open-composer");
    expect(stdout).toContain("Manage AI agents");
  });

  it("launches the TUI when requested", async () => {
    await new Promise<void>((resolve, reject) => {
      const child = spawn("bun", ["run", cliPath, "tui"], {
        stdio: "pipe",
        timeout: 5000,
      });

      let _stdout = "";
      let stderr = "";

      child.stdout?.on("data", (data) => {
        _stdout += data.toString();
      });

      child.stderr?.on("data", (data) => {
        stderr += data.toString();
      });

      setTimeout(() => {
        child.kill("SIGINT");
      }, 1000);

      child.on("close", (code) => {
        expect(code === 0 || code === null).toBe(true);
        const hasRealErrors =
          stderr.length > 0 &&
          !stderr.includes("Warning:") &&
          !stderr.includes("Encountered");
        expect(hasRealErrors).toBe(false);
        resolve();
      });

      child.on("error", (error) => {
        reject(error);
      });
    });
  });

  it("supports gw list", async () => {
    const result = await runCli(["gw", "list"]);
    const stdout = stripAnsi(result.stdout);
    const stderr = stripAnsi(result.stderr);

    expect(result.code).toBe(0);
    expect(stderr).toBe("");
    expect(stdout).toContain("Git worktrees:");
  });
});
