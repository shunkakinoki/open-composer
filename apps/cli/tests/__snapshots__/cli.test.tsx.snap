// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`CLI Execution prints help text that matches snapshot 1`] = `
"Open Composer CLI

USAGE
  $ open-composer <command>

COMMANDS
  agents            Manage AI agents
  pr                Manage GitHub Pull Requests with comprehensive workflow
  gw                Manage git worktrees
  session           Manage persistent process sessions
  sessions          Manage development sessions
  settings          Manage application settings
  spawn             Spawn multiple AI agents in separate worktrees with tmux sessions
  stack             Manage stacked pull requests
  status            Show current status of agents, worktrees, and PRs
  telemetry         Manage telemetry and privacy settings
  tui               Launch the interactive TUI

Run 'open-composer <command> --help' for more information on a specific command.
"
`;
